smart-cards/src/main/java/gov/va/api/health/smartcards/patient/PatientController.java

```diff
@@ -95,28 +130,72 @@ void initDirectFieldAccess(DataBinder dataBinder) {
       @Valid @RequestBody Parameters parameters,
       @RequestHeader(name = "Authorization", required = false) String authorization) {
     checkState(!StringUtils.isEmpty(id), "id is required");
-    var credentialTypes = validateCredentialType(parameters);
+    var credentialTypes = getCredentialTypes(parameters);
+    validateCredentialTypes(credentialTypes);
     Patient.Bundle patients = fhirClient.patientBundle(id, authorization);
     Patient patient = getPatientFromBundle(patients, id);
     Immunization.Bundle immunizations = fhirClient.immunizationBundle(patient, authorization);
+    lookupLocations(immunizations, authorization);
     List<MixedEntry> resources = new ArrayList<>();
     consumeBundle(patients, resources, x -> true, this::transform);
     consumeBundle(immunizations, resources, this::filter, this::transform);
+    // Index unique URLs and replace with 'resource:X' scheme
+    List<String> urls = indexAndReplaceUrls(resources);
     MixedBundle bundle = toBundle(resources);
     var vc = vc(bundle, credentialTypes);
-    var parametersResponse = parameters(vc);
+    var parametersResponse = parameters(vc, urls);
     return ResponseEntity.ok(parametersResponse);
   }
 
+  private void lookupLocations(Immunization.Bundle immunizations, String authorization) {
+    // keep track of locations we already looked up
+    Map<String, Bundle> locations = new HashMap<>();
+    for (Immunization.Entry entry : immunizations.entry()) {
+      Immunization imm = entry.resource();
+      String locationResourceId = Controllers.resourceId(imm.location().reference());
+      Location.Bundle locationBundle = locations.get(locationResourceId);
+      if (locationBundle == null) {
+        locationBundle = fhirClient.locationBundle(locationResourceId, authorization);
```

Comment:
Since we are not using locations the same way as patient/immunization, perhaps it would simplify this code to change the FHIR client to do a Location read and return `Location` instead of `Location.Bundle`?

Reply:
This is true. It used to be Bundle since I needed the entry's `fullUrl`. That's no longer needed though, since we're not adding Location resources to the final VC.

I refactored it so it does a simple read for Location.